src/blockstore/hashmap_db.zig

```diff
@@ -0,0 +1,254 @@
+    }
+
+    pub fn getBytes(
+        self: *Self,
+        comptime cf: ColumnFamily,
+        comptime cf_index: usize,
+        key: cf.Key,
+    ) !?BytesRef {
+        const key_bytes = try cf.key().serializeAlloc(self.allocator, key);
+        defer self.allocator.free(key_bytes);
+        var map = self.maps[cf_index];
+
+        self.transaction_lock.lockShared();
+        defer self.transaction_lock.unlockShared();
+        map.lock.lockShared();
+        defer map.lock.unlockShared();
+
+        const val_bytes = map.getPreLocked(key_bytes) orelse return null;
+
+        const ret = try self.allocator.alloc(u8, val_bytes.len);
+        @memcpy(ret, val_bytes);
+        return .{
+            .allocator = self.allocator,
+            .data = ret,
+        };
+    }
+
+    pub fn delete(
+        self: *Self,
+        comptime cf: ColumnFamily,
+        comptime cf_index: usize,
+        key: cf.Key,
+    ) !void {
+        const key_bytes = try cf.key().serializeAlloc(self.allocator, key);
+        defer self.allocator.free(key_bytes);
+        self.transaction_lock.lockShared();
+        defer self.transaction_lock.unlockShared();
+        _ = self.maps[cf_index].delete(self.allocator, key_bytes);
+    }
+
+    pub fn writeBatch(self: *Self) MapBatch {
+        return .{
+            .allocator = self.allocator,
+            .instructions = .{},
+        };
+    }
+
+    pub fn initBatch(self: *Self) MapBatch {
+        return MapBatch.init(self.allocator);
+    }
+
+    /// Atomicity may be violated if there is insufficient
+    /// memory to complete a PUT.
+    pub fn commit(self: *Self, batch: MapBatch) !void {
+        self.transaction_lock.lock();
+        defer self.transaction_lock.unlock();
+
+        for (batch.instructions.items) |ix| {
+            switch (ix) {
+                .put => |put_ix| {
+                    const cf_index, const key, const value = put_ix;
+                    try self.maps[cf_index].put(key, value);
+                },
+                .delete => |delete_ix| {
+                    const cf_index, const key = delete_ix;
+                    self.maps[cf_index].delete(batch.allocator, key);
+                },
+            }
+        }
+    }
+};
+
+pub const MapBatch = struct {
+    allocator: Allocator,
+    instructions: std.ArrayListUnmanaged(Instruction),
+
+    const Instruction = union(enum) {
+        put: struct { usize, []const u8, []const u8 },
+        delete: struct { usize, []const u8 },
+    };
+
+    const Self = @This();
+
+    pub fn init(allocator: Allocator) Self {
+        return .{
+            .allocator = allocator,
+            .instructions = std.ArrayListUnmanaged(Instruction){},
+        };
+    }
+
+    pub fn deinit(self: Self) void {
+        self.instructions.deinit(self.allocator);
+    }
+
+    pub fn put(
+        self: *Self,
+        comptime cf: ColumnFamily,
+        comptime cf_index: usize,
```

Comment:
Is there a reason this is `comptime`? Does the interface expect it to be, or is this part of some future work?

Reply:
Yeah, this doesn't need to be comptime here. I should probably change this.

The idea is that I want `Database.put` to calculate the index at comptime. I must have put this here as an assertion that Database is actually doing this. But I can see how it is redundant, confusing, and doesn't help with separation of concerns.

Reply:
25f37dbdd8f25ac29873e4df59da491984b22ed1